### PR TITLE
Adding Qsim confusion matrix runtime exception

### PR DESCRIPTION
**Pull request by angelo-laskaris**
QSim currently does not support confusion matrices on measurement gates. See https://github.com/quantumlib/Cirq/issues/6305 for more info. While work is done to support confusion matrices surfacing a runtime exception.

### CODE REVIEW

File: qsimcirq/qsim_simulator.py
Comment (95-martin-orion) [inline qsimcirq/qsim_simulator.py:230]: I recommend moving this to `qsim_circuit.py` and calling it in `QSimCircuit.translate_cirq_to_qsim` instead. Every Cirq circuit that gets converted to qsim must go through that method, so putting it there would save having to call it from every simulation method in this file.

File: qsimcirq/qsim_simulator.py
Comment (95-martin-orion) [inline qsimcirq/qsim_simulator.py:249]: Prefer `ValueError`, since this occurs when a value (`circuit`) is invalid.

File: qsimcirq/qsim_simulator.py
Comment (95-martin-orion) [inline qsimcirq/qsim_simulator.py:246]: Drop `list` - you can iterate over the output of `values()` directly.

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]

File: qsimcirq/qsim_simulator.py
Comment (angelo-laskaris) [reply qsimcirq/qsim_simulator.py]: I moved the method to QSimCircuit and updated the unit test accordingly. However, I want to note the  invocation pattern seen here: https://github.com/quantumlib/Cirq/issues/6305 does not result in an exception being thrown.

**Comment by angelo-laskaris** [comment]
Adding black formatter to my IDE, rebasing and then posting a commit on top of this one

File: qsimcirq/qsim_simulator.py
Comment (95-martin-orion) [reply qsimcirq/qsim_simulator.py]: > the invocation pattern seen here: [quantumlib/Cirq#6305](https://github.com/quantumlib/Cirq/issues/6305) does not result in an exception being thrown.

That's deeply surprising to me. Could you try adding the check in `translate_cirq_to_qtrajectory` as well? https://github.com/quantumlib/qsim/blob/401ebec99628050d6e41cc43925027966f630ebd/qsimcirq/qsim_circuit.py#L438

Comment (angelo-laskaris) [reply qsimcirq/qsim_simulator.py]: I added the check here as well and the `ValueError` still wasn't raised. Your surprise makes me want to make sure my environment doesn't have anything odd/wonky. Let me get a summary of my env (python version qsim ver, cirq ver) and follow up

Comment (angelo-laskaris) [reply qsimcirq/qsim_simulator.py]: Python version: 3.9
-----
cirq==1.2.0
qsimcirq==0.16.3

**Comment by angelo-laskaris** [comment]
@95-martin-orion The one test that failed doesn't seem to be related to the change. Is this test ever flaky?

**Comment by 95-martin-orion** [comment]
Opened #632 to track the issue.

**Comment by 95-martin-orion** [comment]
#632 is fixed - I'm going to go ahead with the v0.17.0 release to unblock some things, but I'm happy to cut a v0.17.1 once this is merged.

**Comment by angelo-laskaris** [comment]
> #632 is fixed - I'm going to go ahead with the v0.17.0 release to unblock some things, but I'm happy to cut a v0.17.1 once this is merged.

Okay sounds good!

**Comment by tanujkhattar** [comment]
@95-martin-orion Are there any further concerns or can we LGTM?

File: qsimcirq/qsim_simulator.py
Comment (95-martin-orion) [reply qsimcirq/qsim_simulator.py]: How was this resolved?

**Comment by 95-martin-orion** [comment]
> @95-martin-orion Are there any further concerns or can we LGTM?

Sorry about the delay - this dropped off my radar. I've unresolved one comment above that I'd like clarity on, otherwise this is good to go. I'll follow up with a release shortly after.

File: qsimcirq/qsim_simulator.py
Comment (angelo-laskaris) [reply qsimcirq/qsim_simulator.py]: I confirmed that there was nothing wonky with my setup or library versions and the indeed in neither `translate_cirq_to_qsim` or `translate_cirq_to_qtrajectory` when the circuit is executed like so

`qsimcirq.QSimSimulator(seed=0).run(circuit, repetitions=100)`

**Comment by angelo-laskaris** [comment]
> > @95-martin-orion Are there any further concerns or can we LGTM?
> 
> Sorry about the delay - this dropped off my radar. I've unresolved one comment above that I'd like clarity on, otherwise this is good to go. I'll follow up with a release shortly after.

No worries! Followed up on your comment

**Review by 95-martin-orion — APPROVED** [review]

**Pull request merged**
